File: src/main/java/net/lomeli/trophyslots/core/criterion/UnlockSlotTrigger.java
package net.lomeli.trophyslots.core.criterion;

import com.google.gson.JsonObject;
import net.lomeli.trophyslots.TrophySlots;
import net.lomeli.trophyslots.core.capabilities.IPlayerSlots;
import net.lomeli.trophyslots.core.capabilities.PlayerSlotHelper;
import net.minecraft.advancements.critereon.AbstractCriterionTriggerInstance;
import net.minecraft.advancements.critereon.DeserializationContext;
import net.minecraft.advancements.critereon.EntityPredicate;
import net.minecraft.advancements.critereon.SimpleCriterionTrigger;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;

public class UnlockSlotTrigger extends SimpleCriterionTrigger<UnlockSlotTrigger.Instance> {
    private static final String CRITERION_ID = "unlock_slot";
    private static final String COUNT_JSON_KEY = "slots_unlocked";

    @Override
    @SuppressWarnings("NullableProblems")
    protected Instance createInstance(JsonObject json, EntityPredicate.Composite comp, DeserializationContext context) {
        return new Instance(getId(), comp, json.get(COUNT_JSON_KEY).getAsInt());
    }

    @Override
    @SuppressWarnings("NullableProblems")
    public ResourceLocation getId() {
        return new ResourceLocation(TrophySlots.MOD_ID, CRITERION_ID);
    }

    public void trigger(ServerPlayer player) {
        this.trigger(player, (instance -> instance.test(player)));
    }

    static class Instance extends AbstractCriterionTriggerInstance {
        private final int minSlot;

        Instance(ResourceLocation id, EntityPredicate.Composite comp, int minSlot) {
            super(id, comp);
            this.minSlot = minSlot;
        }

        boolean test(ServerPlayer player) {
            IPlayerSlots playerSlots = PlayerSlotHelper.getPlayerSlots(player);
            if (playerSlots != null)
                return minSlot == -1 ? playerSlots.maxSlotsUnlocked() : playerSlots.getSlotsUnlocked() >= minSlot;
            return false;
        }
    }
}

File: src/main/kotlin/net/lomeli/trophyslots/core/triggers/UnlockSlotTrigger.kt
package net.lomeli.trophyslots.core.triggers

import com.google.common.collect.Lists
import com.google.common.collect.Maps
import com.google.common.collect.Sets
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonObject
import net.lomeli.trophyslots.TrophySlots
import net.lomeli.trophyslots.capabilities.slots.SlotManager
import net.minecraft.advancements.ICriterionTrigger
import net.minecraft.advancements.PlayerAdvancements
import net.minecraft.advancements.critereon.AbstractCriterionInstance
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.util.ResourceLocation

class UnlockSlotTrigger : ICriterionTrigger<UnlockSlotTrigger.Instance> {
    private val listeners = Maps.newHashMap<PlayerAdvancements, UnlockSlotTrigger.Listeners>()

    override fun getId(): ResourceLocation = ResourceLocation(TrophySlots.MOD_ID, "unlock_slot")

    override fun removeAllListeners(playerAdvancementsIn: PlayerAdvancements?) {
        this.listeners.remove(playerAdvancementsIn)
    }

    override fun deserializeInstance(json: JsonObject?, context: JsonDeserializationContext?): Instance {
        return Instance(json?.get("slots_unlocked")?.asInt!!)
    }

    override fun removeListener(playerAdvancementsIn: PlayerAdvancements?, listener: ICriterionTrigger.Listener<Instance>?) {
        var trigger: UnlockSlotTrigger.Listeners? = this.listeners[playerAdvancementsIn] ?: return
        trigger!!.remove(listener!!)
        if (trigger.isEmpty) this.listeners.remove(playerAdvancementsIn)
    }

    override fun addListener(playerAdvancementsIn: PlayerAdvancements?, listener: ICriterionTrigger.Listener<Instance>?) {
        var trigger: UnlockSlotTrigger.Listeners? = this.listeners[playerAdvancementsIn]

        if (trigger == null) {
            trigger = UnlockSlotTrigger.Listeners(playerAdvancementsIn!!)
            this.listeners.put(playerAdvancementsIn, trigger)
        }

        trigger.add(listener!!)
    }

    fun trigger(player: EntityPlayerMP) {
        var listenter = this.listeners[player.advancements]
        if (listenter != null) listenter.trigger(player)
    }

    class Instance(private val minSlot: Int) : AbstractCriterionInstance(ResourceLocation(TrophySlots.MOD_ID, "unlock_slot")) {
        fun test(player: EntityPlayerMP): Boolean {
            val slotInfo = SlotManager.getPlayerSlotInfo(player)!!
            return if (minSlot == -1) slotInfo.isAtMaxSlots() else slotInfo.getSlotsUnlocked() >= minSlot
        }
    }

    internal class Listeners(private val playerAdvancements: PlayerAdvancements) {
        private val listeners = Sets.newHashSet<ICriterionTrigger.Listener<UnlockSlotTrigger.Instance>>()

        val isEmpty: Boolean
            get() = this.listeners.isEmpty()

        fun add(listener: ICriterionTrigger.Listener<UnlockSlotTrigger.Instance>) {
            this.listeners.add(listener)
        }

        fun remove(listener: ICriterionTrigger.Listener<UnlockSlotTrigger.Instance>) {
            this.listeners.remove(listener)
        }

        fun trigger(player: EntityPlayerMP) {
            var list: MutableList<ICriterionTrigger.Listener<UnlockSlotTrigger.Instance>>? = null

            for (listener in this.listeners) {
                if ((listener.criterionInstance as UnlockSlotTrigger.Instance).test(player)) {
                    if (list == null)
                        list = Lists.newArrayList<ICriterionTrigger.Listener<UnlockSlotTrigger.Instance>>()

                    list!!.add(listener)
                }
            }

            if (list != null) {
                for (listener1 in list)
                    listener1.grantCriterion(this.playerAdvancements)
            }
        }
    }
}